File: BaekJoon/src/graph/BaekJoon_1325.java
package graph;

import java.io.*;
import java.util.*;

/**
 * BaekJoon_1325, 효율적인 해킹 
 * 	
 * @author kevin-Arpe
 * 
 * Sketch Idea
 * 	1. 모든 노드를 순회하며 BFS 탐색 
 * 	2. 
 *
 */
public class BaekJoon_1325 {
	static int N, M, max;
	static List<Integer>[] list;
	static int[] cnt;
	
	static void bfs(int start) {
		Queue<Integer> q = new LinkedList<>();
		boolean[] isVisited = new boolean[N + 1];
		q.offer(start);
		isVisited[start] = true;
		
		int count = 0;
		while (q.size() > 0) {
			int curr = q.poll();
			for (int next : list[curr]) {
				if (isVisited[next]) continue;
				q.add(next);
				isVisited[next] = true;
				count++;
			}
		}
		cnt[start] = count;
		max = Math.max(max, cnt[start]);
	}

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		
		list = new ArrayList[N + 1];
		for (int i=1; i<=N; i++) {
			list[i] = new ArrayList<>();
		}
		for (int i=0; i<M; i++) {
			st = new StringTokenizer(br.readLine());
			int A = Integer.parseInt(st.nextToken());
			int B = Integer.parseInt(st.nextToken());
			list[B].add(A);
		}
		
		cnt = new int[N + 1];
		for (int i=1; i<=N; i++) {
			bfs(i);
		}
		
		StringBuilder sb = new StringBuilder();
		for (int i=1; i<=N; i++) {
			if (cnt[i] == max) sb.append(i).append(" ");
		}
		System.out.println(sb.toString().trim());
		br.close();
	}

}

File: BaekJoon/src/graph/BaekJoon_1325.js
/**
 * BaekJoon_1325, 효율적인 해킹
 *  - 문제 분류: 그래프 탐색
 */
const solution = () => {
  class Node {
    constructor(idx, depth, start) {
      this.idx = idx;
      this.depth = depth;
      this.start = start;
    }
  }

  /* 변수 관리 */
  const input = require("fs").readFileSync("./dev/stdin/1325.txt").toString().split("\n");
  const [N, M] = input[0].split(" ").map(Number);
  const adjList = Array.from({ length: N + 1 }, () => [[], []]);
  let v = null;
  let dp = null;
  let max = 0;
  const list = [];
  const answer = [];

  /* 메인 로직 */
  for (let i = 1; i <= M; i++) {
    const [from, to] = input[i].split(" ").map(Number);
    adjList[from][0].push(to);
    adjList[to][1].push(from);
  }

  for (let i = 1; i <= N; i++) {
    if (adjList[i][0].length > 0) continue;
    v = Array.from({ length: N + 1 }, () => false);
    dp = Array.from({ length: N + 1 }, () => 0);
    v[i] = true;
    const count = dfs(i, i);

    if (count > max) max = count;
    list.push([i, count]);
  }

  for (const [idx, count] of list) {
    if (count === max) answer.push(idx);
  }

  /* 정답 반환 */
  return answer.join(" ");

  // bfs
  function dfs(node, start) {
    let count = 1;

    for (const next of adjList[node][1]) {
      if (v[next]) continue;
      if (dp[next] > 0) {
        count += dp[next];
        continue;
      }

      // console.log(node, "에서 ", next, "로 감");
      v[next] = true;
      count += dfs(next, start);
    }

    dp[node] = count;
    return count;
  }
};

console.log(solution());
